### PR TITLE
txpool to pull blocks of transactions until no gas remains for the block

### PR DESCRIPTION
**Pull request by hexoscott**
Moving the txpool transaction pull into the execution phase and looping until the gas is used up or the txpool runs dry.

Removed the concept of local vs remote transactions as comments/code showed this split was no longer in use.  

Created a `PreparedTxs` collection to satisfy the use case in the integration tool when mining is active.

Untested locally as I have no way of mining/validating currently.

### CODE REVIEW

**Review by AskAlexSharov — APPROVED** [review]

File: go.mod [hunk 1]
```diff
@@ -257,3 +257,5 @@ require (
 )
 
 replace github.com/tendermint/tendermint => github.com/bnb-chain/tendermint v0.31.12
+
+replace github.com/ledgerwatch/erigon-lib => github.com/hexoscott/erigon-lib v0.0.0-20221130104614-9b44654b2826
```
Comment (AskAlexSharov) [inline go.mod:261]: likely need remove it

**Review by AskAlexSharov — CHANGES_REQUESTED** [review]

File: go.mod (hunk 1 above)
Comment (AskAlexSharov) [reply go.mod:261]: granted you access to create PR's in erigon-lib repo

**Review by AskAlexSharov — CHANGES_REQUESTED** [review]

File: eth/stagedsync/stage_mining_exec.go [hunk 2]
```diff
@@ -176,6 +211,62 @@ func SpawnMiningExecStage(s *StageState, tx kv.RwTx, cfg MiningExecCfg, quit <-c
...
+		s := rlp.NewStream(bytes.NewReader(txSlots.Txs[i]), uint64(len(txSlots.Txs[i])))
+
+		transaction, err := types.DecodeTransaction(s)
+		if err == io.EOF {
+			continue
+		}
+		if err != nil {
+			return nil, err
+		}
+		if transaction.GetChainID().ToBig().Uint64() != 0 && transaction.GetChainID().ToBig().Cmp(cfg.chainConfig.ChainID) != 0 {
```
Comment (AskAlexSharov) [inline eth/stagedsync/stage_mining_exec.go:250]: u256 has method .IsZero

Comment (AskAlexSharov) [reply eth/stagedsync/stage_mining_exec.go:250]: I think we can cast ChainId to uint256 at miner startup

File: eth/stagedsync/stage_mining_exec.go [hunk 3]
```diff
@@ -176,6 +211,62 @@ func SpawnMiningExecStage(s *StageState, tx kv.RwTx, cfg MiningExecCfg, quit <-c
...
+			continue
+		}
+
+		var sender common.Address
+		copy(sender[:], txSlots.Senders.At(i))
```
Comment (AskAlexSharov) [inline eth/stagedsync/stage_mining_exec.go:255]: in-theory can cast next way: https://tip.golang.org/doc/go1.17#language

File: go.mod (hunk 1 above)
Comment (hexoscott) [reply go.mod:261]: There's a PR open already there for the changes to the txpool `Best` function

File: eth/stagedsync/stage_mining_exec.go (hunk 3 above)
Comment (hexoscott) [reply eth/stagedsync/stage_mining_exec.go:255]: using unsafe.Slice?

Comment (AskAlexSharov) [reply eth/stagedsync/stage_mining_exec.go:255]: no "An expression s of type []T may now be converted to array pointer type *[N]T"

File: eth/stagedsync/stage_mining_exec.go (hunk 2 above)
Comment (hexoscott) [reply eth/stagedsync/stage_mining_exec.go]: Both addressed now

File: go.mod (hunk 1 above)
Comment (hexoscott) [reply go.mod]: The changes in erigon-lib are merged now so this is updated to the actual version from main now

**Pull request merged**